tsdb/engine/tsm1/engine.go

```diff
@@ -993,6 +975,100 @@ func (e *Engine) createVarRefIterator(opt influxql.IteratorOptions) ([]influxql.
 	return itrs, nil
 }
 
+// createTagSetIterators creates a set of iterators for a tagset.
+func (e *Engine) createTagSetIterators(ref *influxql.VarRef, mm *tsdb.Measurement, t *influxql.TagSet, opt influxql.IteratorOptions) ([]influxql.Iterator, error) {
+	// Set parallelism by number of logical cpus.
+	parallelism := runtime.GOMAXPROCS(0)
+	if parallelism > len(t.SeriesKeys) {
+		parallelism = len(t.SeriesKeys)
+	}
+
+	// Create series key groupings w/ return error.
+	groups := make([]struct {
+		keys    []string
+		filters []influxql.Expr
+		itrs    []influxql.Iterator
+		err     error
+	}, parallelism)
+
+	// Group series keys.
+	n := len(t.SeriesKeys) / parallelism
+	for i := 0; i < parallelism; i++ {
+		group := &groups[i]
+
+		if i < parallelism-1 {
+			group.keys = t.SeriesKeys[i*n : (i+1)*n]
+			group.filters = t.Filters[i*n : (i+1)*n]
+		} else {
+			group.keys = t.SeriesKeys[i*n:]
+			group.filters = t.Filters[i*n:]
+		}
+
+		group.itrs = make([]influxql.Iterator, 0, len(group.keys))
+	}
+
+	// Read series groups in parallel.
+	var wg sync.WaitGroup
+	for i := range groups {
+		wg.Add(1)
+		go func(i int) {
+			defer wg.Done()
+			groups[i].itrs, groups[i].err = e.createTagSetGroupIterators(ref, mm, groups[i].keys, t, groups[i].filters, opt)
+		}(i)
+	}
+	wg.Wait()
+
+	// Determine total number of iterators so we can allocate only once.
+	var itrN int
+	for _, group := range groups {
+		itrN += len(group.itrs)
+	}
+
+	// Combine all iterators together and check for errors.
+	var err error
+	itrs := make([]influxql.Iterator, 0, itrN)
+	for _, group := range groups {
+		if group.err != nil {
+			err = group.err
```

Comment:
They all need to get closed if there is an error. I could move to a separate loop to close each group below but this seemed easier since I had to collect them anyway.